### PR TITLE
feat(relational_iter): add state_table_iter with  pk_prefix and pk_bounds

### PR DESCRIPTION
**Pull request by wcy-fdu**
## What's changed and what's your intention?

This PR adds `with_prefix` and `with_pk_bounds` for `StateTableIter`, the former can be used while refactoring join with relational table. 
For example,  a table with pk
`(a, b)`
`(1, 1)`
`(1, 2)`
`(2, 1)`
`(2, 2)`
`(3, 1)`
`(3, 2)`
iter(a = 2) will scan `(2, 1), (2, 2)`, iter(a > 2) will scan `(2, 1), (2, 2), (3, 1), (3, 2)`.
## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
close https://github.com/singularity-data/risingwave/issues/2811

### CODE REVIEW

File: src/storage/src/table/state_table.rs [hunk 1]
```diff
@@ -135,6 +136,22 @@ impl<S: StateStore> StateTable<S> {
             epoch,
         ))
     }
+
+    pub async fn iter_with_bounds(
+        &self,
+        pk_bounds: Range<Row>,
```
Comment (xxchan) [inline src/storage/src/table/state_table.rs:142]: Use `RangeBounds` in APIs (like in `StateStore`) so that we can use unbounded range e.g., `start..`?

Comment (wcy-fdu) [reply src/storage/src/table/state_table.rs:142]: > btw, how to represent a prefix scan? e.g., pk=[a,b], scan a=1

`iter_with_pk_prefix`  has been added, and maybe we can keep these two interfaces(`iter_with_pk_prefix` and `iter_with_pk_bound`).

Comment (wcy-fdu) [reply src/storage/src/table/state_table.rs:142]: > btw, how to represent a prefix scan? e.g., pk=[a,b], scan a=1

`iter_with_pk_prefix` has been added, and maybe we can keep these two interfaces(`iter_with_pk_prefix` and `iter_with_pk_bound`).

Comment (xxchan) [reply src/storage/src/table/state_table.rs:142]: Yes, but still need `RangeBounds`, which includes all of these range types
image

File: src/storage/src/table/state_table.rs [hunk 2]
```diff
@@ -235,6 +274,221 @@ impl<S: StateStore> StateTableRowIter<S> {
...
+                    mem_table_iter.next().unwrap()?;
+
+                    unreachable!()
+                }
+            }
+        }
+    }
+
+    #[try_stream(ok = Cow<'a, Row>, error = StorageError)]
+    async fn into_stream_with_prefix<'a>(
```
Comment (fuyufjh) [inline src/storage/src/table/state_table.rs:393]: Looks a little duplicated with `into_stream_with_bounds`. Is it possible to unify them into one? 🤔 (Prefix is also some kind of ranges, I think)

**Review by fuyufjh — APPROVED** [review]
Generally LGTM

File: src/storage/src/table/state_table.rs (hunk 2 above)
Comment (wcy-fdu) [reply src/storage/src/table/state_table.rs]: Thanks for your comments and I have simplify the duplicated part.

File: src/storage/src/table/state_table.rs [hunk 3]
```diff
@@ -135,6 +138,96 @@ impl<S: StateStore> StateTable<S> {
...
+        pk_prefix: Row,
+        prefix_serializer: OrderedRowSerializer,
+        epoch: u64,
+    ) -> StorageResult<impl RowStream<'_>> {
+        let mem_table_iter = self.mem_table.buffer.iter();
+        let key_bytes = serialize_pk(&pk_prefix, &prefix_serializer).map_err(err)?;
+        let start_key_with_prefix = self.keyspace.prefixed_key(&key_bytes);
+        let cell_based_bounds = (
+            Included(start_key_with_prefix.clone()),
+            Included(next_key(start_key_with_prefix.as_slice())),
```
Comment (xxchan) [inline src/storage/src/table/state_table.rs:215]: next key should be excluded?

**Pull request merged**